Write the EDO chapter (#66)
daffidwilde/thesis: chapters/edo/main.tex

simulated) benchmark datasets based on some fixed metrics. The algorithm(s) with
the smallest value of this metric are chosen to be the `best performing'. We
offer a new approach to flip this paradigm. We instead aim to gain a richer
picture of the performance of an algorithm by generating artificial data through
genetic evolution, the purpose of which is to create populations of datasets for
which a particular algorithm performs well on a given metric. These datasets can
be studied so as to learn what attributes lead to a particular progression of a
given algorithm.  Following a detailed description of the algorithm as well as a
brief description of an open source implementation, a case study in clustering
is presented. This case study demonstrates the performance and nuances of the
method which we call Evolutionary Dataset Optimisation. In this study, a number
of known properties about preferable datasets for the clustering algorithms
known as \(k\)-means and DBSCAN are realised in the generated datasets.

\myrule%

The differences between this chapter and the publication are an extended
discussion of the motivation behind the Evolutionary Dataset Optimisation method

> and its components, as well as a revised case study which concludes the chapter.

Author: Fixed 392ae08

                \left(\mathbb{N}\cup\{\infty\}\right)
            }^2~|~c_{\min} \leq c_{\max}\right\}
        \]
        for each \(j = 1, \ldots, |\mathcal{P}|\). That is, \(C\) defines the
        minimum and maximum number of columns a dataset may have from each
        distribution in \(\mathcal{P}\).
    \item A second selection parameter, \(l \in [0, 1]\), to allow for a
        small proportion of `lucky' individuals to be carried forward.
    \item A shrink factor, \(s \in [0, 1]\), defining the relative size of a
        component of the search space to be retained after adjustment.
\end{itemize}

This chapter discusses the components and mechanisms of the EDO method in a
largely mathematical manner. However, a Python package,
\href{https://github.com/daffidwilde/edo}{\mintinline{python}{edo}}, that
implements the EDO algorithm is used to demonstrate its practical use and
particular technical aspects of the method. This implementation is built on the
scientific Python stack~\cite{pandas,numpy} and has been developed to be
consistent with the current best practices of open source software
development~\cite{Jimenez2017} so that it is modular, automatically tested and

> fully documented. It is freely available online under the MIT

Author: Fixed c9fd369

> licence~\cite{edo-project}.

Author: Fixed c9fd369

with the mutation probability, \(p_m\). Following subsequent testing, and in the
process of documenting the Python library, it was decided that allowing for this
kind of mutation made for confusing results. In particular, studying the
resultant individuals became more complicated when individuals retained values
in their columns that were now beyond any reasonable bounds of the associated
distribution, for instance. Since removing this stage of the process, no
noticeable impact has been identified on the ability of the EA to traverse the
search space compared with its inclusion.

\inputtikz{mutation}{The mutation process}

Each of the potential mutations occurs with the same probability \(p_m\).
However, how the columns are maintained assure that (assuming appropriate
choices for \(f\) and \(\mathcal{P}\)) even multiple mutations in the dataset
will only result in some incremental change in the individual's fitness relative
to, say, a completely new individual.

\inputalg{mutation}

\section{A case study in clustering}\label{section:examples}

The following case study contains three examples that act as a form of

> validation for EDO. These examples also highlight some of the nuances in its

Author: Fixed by 024815b

find datasets that minimise the final inertia found by \(k\)-means clustering.

% TODO include a list of known results dressed as `hypotheses'?

For visualisation purposes, these examples will restrict EDO to datasets that
are two-dimensional,~i.e. \(C = ((2, 2))\). Then, for simplicity, each dataset
will be clustered into three parts,~i.e. \(k = 3\), and have its columns formed
from uniform distributions, denoted by \(U\), enclosed by the unit interval.
Thus, the search space is the unit square, and the only element of
\(\mathcal{P}\) is:
\begin{equation}\label{eq:uniform}
    \mathcal{U} := \left\{U(\alpha, \beta)~|~\alpha, \beta \in [0, 1]\right\}
\end{equation}

The remaining parameters are as follows: \(N=100\), \(R=(50,100)\), \(M=100\),
\(b=0.1\), \(l=0\), \(p_m=0.01\), and shrinkage is excluded. This set of
parameters has been adapted from that used in~\cite{Wilde2020:edo}. The changes
are: omitting the trivial case where the number of rows equals \(k\); shortening
the run time to reduce computational resources, and because the EA produces as
useful results, indicating the identification of a local optimum; and, finally,

> increasing the selective pressure (by reducing \(b\)) to mitigate the effect of

Reviewer: I don't fully understand "and because the EA produces as useful results" here?

    across the given number of trials with k=3. """

    data = MinMaxScaler().fit_transform(individual.dataframe, copy=True)

    inertias = []
    for seed in range(max_seeds):
        km = KMeans(n_clusters=3, random_state=seed).fit(data)
        inertias.append(km.inertia_)

    return min(inertias)

size = 100
row_limits = [50, 100]
col_limits = [2, 2]
max_iter = 100
best_prop = 0.1
lucky_prop = 0
mutation_prob = 0.01

Uniform.param_limits["bounds"] = [0, 1]
distributions = [Uniform]
\end{sourcepy}
\caption{%
    An abridged version of the experiment configuration script used in the first
    example
}\label{snp:inertia}
\end{listing}

\begin{listing}
\begin{usagesh}
> cd /path/to/experiments
> edolab run --seeds=10 --cores=4 kmeans_inertia.py
> edolab summarise --tarball kmeans_inertia.py
\end{usagesh}
\caption{Example usage of the \mintinline{python}{edolab} command-line tool}
\label{snp:edolab}
\end{listing}

Once the EDO algorithm has terminated, a body of datasets, and information about
those datasets, is recorded. This output is referred to as a \emph{history}.
Figure~\ref{fig:inertia_progression} shows the progression of the fitness

> function (inertia) and the number of rows at ten generation intervals across the

Reviewer: Just a thought, not sure if here is the right place but is it worth having a sentence about the pure volume of data that eco can produce? (Not necessarily a good/bad thing but worth being aware of).

\begin{figure}
    \centering
    \includegraphics[width=\linewidth]{kmeans_inertia_inds.pdf}
    \caption{%
        Representative individuals from EDO trials with inertia
    }\label{fig:kmeans_inertia_inds}
\end{figure}

In this case, as may have been expected, the worst individuals take the form of
random clouds with no distinct clusters. However, there are some patterns in the
better-performing individuals. It is clear that there is a preference for tight
clusters, but also it appears that well-performing datasets have columns with a
strong positive correlation. Having such a relationship may seem irrelevant to
the success of \(k\) means, but in doing so, the dataset becomes
one-dimensional. Removing a dimension reduces the search space of the algorithm
considerably, and makes it easier for the \(k\)-means algorithm to achieve its
real goal of finding the centroidal Voronoi tessellation of a
dataset~\cite{Du2006}. When restricted to a single dimension and with \(k = 3\),
the optimal tessellation is equivalent to finding a clustering that matches the
tertiles of a set of numbers, and when \(k = 2\) this is the same as finding the
median.

> % TODO a plot of difference between the cell boundaries and the true tertiles?

Reviewer: Commenting so this isn't forgotten.

In this way, inertia could be considered a flawed fitness function. Without the
scaling step, for instance, inertia produces near-trivial results, which begs
the question: is there anything else to be learnt here? The answer is,
``probably''. There are two options available to draw more learning from this
algorithm. Either change the parameters passed to EDO, or modify the fitness
function. Given that useful results have already been found with this parameter
set, and to avoid cherry-picking any further results by tweaking parameters,
this study considers the latter for the remaining examples.

\subsection{The silhouette coefficient}\label{subsec:silhouette}

In the example above, the presence of strong positive correlations stood out
when using inertia as the fitness function --- as such, counteracting that
effect is the focus of this example.

This study aims to understand \(k\)-means clustering, and therefore, the fitness
function(s) used should somehow measure the efficacy of the identified
clustering. The silhouette coefficient is one such metric. The silhouette
coefficient evaluates what is colloquially referred to as the `appropriateness'

> of a particular clustering of a dataset. The metric considers both the

Reviewer: Add a reference here?

Author: Fixed ef035bb

the exercise pointless. However, its effect integrates well with the silhouette
coefficient:
\begin{itemize}
    \item The optimal score is the same (\(1 - 0 = 1\)) as the silhouette
        fitness function while the worst score is similar (\(0-1=-1\)).
    \item A score of zero still indicates there is little to be gained in that
        (at the extremes) the dataset has a perfect silhouette and perfect
        correlation or no silhouette with no correlation, indicating a random
        cloud.
    \item Negative scores indicate a dataset with a low silhouette coefficient
        and high correlation,~i.e.\ high inertia but correlated and, therefore,
        unwanted.
\end{itemize}

Figures~\ref{fig:discounted_progression}~and~\ref{fig:kmeans_discounted_inds}
show a summary of the results generated using this discounted silhouette
function with the same parameters as used in the previous examples. The fitness
progression shows a steady increase for the best individuals at each generation
--- as has been the case with the first two cases --- and there is the same
preference for datasets with fewer rows. This trend in the fitness means that

> EDO is certainly doing its job of optimising across the search space. However,

Reviewer: ```suggestion
EDO is indeed optimising across the search space. However,
```

there appears to be some variation in the population fitness here, which may
indicate that the parameter set requires some tweaking or, simply, that the

> environment in which individuals exist is more competitive in a sense. Since the

Reviewer: ```suggestion
environment in which individuals exist is more competitive. Since the
```

roughly equal size and are convex. Without this condition, up to the correct
choice of \(k\), the algorithm will fail to produce satisfactory results for
either inertia or silhouette. This condition is derived from the link between
\(k\)-means and Voronoi tessellation.~\cite{Sonka1993} defines one measure of
the convexity of a set of points (such as a cluster), denoted \(\mathcal C\), is
the ratio of the areas of its concave and convex hulls, denoted \(H_c\) and
\(H_v\), respectively:
\begin{equation}
    \mathcal C :=
    \frac{\text{area}\left(H_c\right)}{\text{area}\left(H_v\right)}
\end{equation}

With this definition, it should be clear that a perfectly convex cluster, such
as a single point, line or convex polygon, would have \(\mathcal{C} = 1\). Also,
it appears that the mean convexity of the clustering increases as fitness
increase --- save for the `invalid' individual --- and suggests that this
condition for convex clusters is sought out during the optimisation process.

% TODO Perhaps a scatter of fitness and convexity for the parents.
% TODO A hypothesis test even? Depends on the scatter, I guess.

> % TODO Should the cluster statistics just be printed in tables for each example?

Reviewer: Perhaps all of the above?

> % TODO Conclude the subsection.

Reviewer: Making sure these aren't missed.

such metric.

The standard silhouette function is used here over that defined
in~\eqref{eq:discounted_silhouette} for two reasons. First, the relationship
between correlation and DBSCAN has not yet been established in this study, and
second, to simplify the final fitness function. An adjustment
to~\eqref{eq:silhouette} must be made since the silhouette coefficient requires
at least two clusters and DBSCAN need only cluster a subset of a dataset
(referred to as the \emph{core~points}), labelling the remainder as outliers.
Let \(S_k (X)\) and \(S_D (X)\) denote the silhouette coefficients of the
clustering found by \(k\)-means and DBSCAN respectively, and let \(Z_D\) be the
clustering found by DBSCAN. Then the \emph{\(k\)-means-preferable} fitness
function is defined to be:
\begin{equation}\label{eq:kmeans_preferable}
    f(X) :=
        \begin{cases}
            S_k (X) - S_D (X), &\quad \text{%
                \begin{tabular}{l}%
                    if \(\left|Z_D\right| > 1\)
                \end{tabular}
            }\\
            - \infty &\quad \ \ \text{otherwise}
        \end{cases}
\end{equation}

> % TODO include a statement of DBSCAN? It's quite long 

Reviewer: I think it's ok to just point at reference text...

Author: Fixed ef035bb

shows the representative individuals for this example from worst to best. Each
dataset is shown with its clustering (and associated silhouette) by
(\subref{fig:kmeans_preferable_kmeans_inds})~\(k\)-means and
(\subref{fig:kmeans_preferable_dbscan_inds})~DBSCAN. In addition to a scattering
of the cluster's data points, the latter figure displays the concave and convex
hulls of each cluster using shading and outline, respectively. A cluster's
\emph{concave hull} is taken to be the \(\alpha\)-shape of the cluster's data
points~\cite{Edelsbrunner1983} where \(\alpha \in \mathbb R\) is the smallest
value such that all points in the cluster are contained in a single polygon.

The best-performing individual, when clustered by \(k\)-means, shows three
distinct clusters, one of which is nicely separated from the others. The
dimension-collapsing effect of positive correlation has come into play for the
two closer clusters, although it has not dominated. In contrast, when DBSCAN
clusters the same dataset, the method identifies three clusters of core points
that exist within the convex hulls of one another, meaning there are overlapping
clusters and, hence, a negative silhouette coefficient.

> % TODO Something about convexity

Reviewer: Making sure this isn't missed.

Author: Fixed by bee19d5

hotspots with impunity. Then, as fitness improves, the clusters become smaller,
contain fewer points and are further apart. All of these factors serve to
maximise the silhouette of the DBSCAN clustering while leaving that of
\(k\)-means mostly unchanged. In general, being able to identify outliers is one
of DBSCAN's core strengths. However, this success is so reliant on being able to
ignore the majority of the data points and finding more delicate cases for the
success of DBSCAN should adapt the fitness function to mitigate this behaviour,
perhaps with a penalty.  

As has already been stated, the software implementation of the EDO method has
been produced in line with the best practices of open source software
development and reproducible research. In aid of this, all the source code used
in these examples (including to create the figures) has been archived under the
DOI \href{https://doi.org/10.5281/zenodo.3492236}{10.5281/zenodo.3492236}.
Likewise, all the data produced to support this case study have been archived
under the DOI
\href{https://doi.org/10.5281/zenodo.3492228}{10.5281/zenodo.3492228}.
% TODO Update these DOIs.

> % TODO Move this paragraph up to the top of the section.

Reviewer: Possibly the top of the chapter.

Author: Fixed by f6ebf9f and 331b92b

accordingly. The case study in this chapter examined several iterations of
fitness functions, but each was adjusted according to what was apparent at the
time. These adjustments were possible because of the architecture of the
implementation of this method. A similar strategy could be employed
automatically by a more sophisticated fitness function that retains some
information about the datasets generated from previous runs of EDO on a
particular (or at least similar) parameter set. In this way, the currently
completely unsupervised learning conducted by the EA could be ushered away from
less helpful solutions (via some penalty, say) and toward previously unexplored
behaviours. This automatic, iterative application of the proposed method would
likely reveal more sophisticated insights into a particular algorithm.

In essence, the EDO method is merely a tool that demonstrates the benefit of the
flipped paradigm set out in this work. The concept of where `good' datasets
exist is not well-documented in the literature, and this thesis hopes that
Evolutionary Dataset Optimisation acts as a starting point for further works to

> come.

Reviewer: As mentioned above, somewhere in this section we need to point at the matching chapter.

Author: Fixed by 024815b
